Fix for Unicode decoding issue when using incremental networking.

LeoNatan opened the pull request:
This is **a critical issue**.

The issue arises when incremental networking is enabled from JS by setting `onprogress` or `onload` on an `XMLHttpRequest` object.

The results:

![example1](https://cloud.githubusercontent.com/assets/2270433/18829964/5a54ff30-83e7-11e6-9806-97857dce0430.png)

![example2](https://cloud.githubusercontent.com/assets/2270433/18829966/5bf40a66-83e7-11e6-84e6-9e4d76ba4f8b.png)

Unicode characters get corrupted seemingly in random. The issue is from the way Unicode character parsing is handled in `RCTNetworking.mm`. When incremental networking is enabled, each chunk of data is decoded and passed to JS:

``` objective-c
incrementalDataBlock = ^(NSData *data, int64_t progress, int64_t total) {
NSString *responseString = [RCTNetworking decodeTextData:data fromResponse:task.response];
if (!responseString) {
  RCTLogWarn(@"Received data was not a string, or was not a recognised encoding.");
  return;
}
NSArray<id> *responseJSON = @[task.requestID, responseString, @(progress), @(total)];
  [self sendEventWithName:@"didReceiveNetworkIncrementalData" body:responseJSON];#9104
};
```

However, this is not correct when Unicode characters are longer than one byte per character. The chunk of data is arbitrarily cut off, which may be in the middle of a character. In `decodeTextData:fromResponse:`, there is an attempt to decode the data with the encoding from the response headers, but if this decode fails (most likely due to the issue described here), a fallback is performed by attempting to guess the string encoding. This is a lossy operation and causes issues, as seen in the screenshots above.

The correct behavior should be to take the incorrectly cut off bytes and save them as carry for the next data chunk. After implementing this fix, the Unicode characters are parsed correctly.

As a side effect, because there is no need to guess the encoding anymore, performance is improved significantly in some cases.

See https://github.com/LeoNatan/RN-Unicode-Decode-Woes for a demo project with the proposed fix.

This is a potential fix for
# 7646
# 8914
# 9104

facebook-github-bot-9 commented:
By analyzing the blame information on this pull request, we identified @javache and @philikon to be potential reviewers.

facebook-github-bot-9 commented:
Thank you for your pull request.  As you may know, we require contributors to sign our Contributor License Agreement, and we don't seem to have you on file and listed as active anymore.  In order for us to review and merge your code, please email cla@fb.com with your details so we can update your status.

facebook-github-bot-0 commented:
@LeoNatan updated the pull request - [view changes](https://github.com/facebook/react-native/pull/10110/files/55df46e3b0e801fd2f3c91cd29bb3ce8f515b8d8..6258d5ed416bf746a8719e7d606d913d2584b3bb)

facebook-github-bot-5 commented:
Thank you for signing our Contributor License Agreement. We can now accept your code for this (and any) Facebook open source project. Thanks!

gran33 commented:
👍

Caijiacheng commented:
👍

psrustik commented:
:+1:

kikimor commented:
:+1:

militska commented:
+1

Caijiacheng commented:
hello, When this PR will  merge into RN? this sounds a critical issue.but, it take some long time to fix it?

LeoNatan commented:
@javache @philikon Any news on that review?

lacker commented:
Could you add a test for this behavior?

LeoNatan commented:
@lacker Hi Kevin,

Where would be the best place to put my test? I see there are tests in UIExplorer.

Thanks

lacker commented:
Yeah, I think that's a good place to put iOS-centric tests.

LeoNatan commented:
@lacker Hi Kevin,

I added a test for `decodeTextData:fromResponse:withCarryStorage:`. Testing the entire `RCTNetworking` becomes a very complex task, needing to mock network and syncing the async architecture. I think the test I added is good enough to catch issues in the future.

fkgozali commented on Examples/UIExplorer/UIExplorerUnitTests/RCTUnicodeDecodeTests.m:
```diff
@@ -0,0 +1,46 @@
+//  UIExplorer
+//
+//  Created by Leo Natan (Wix) on 27/10/2016.
+//  Copyright © 2016 Facebook. All rights reserved.
```
@lacker: what's the convention for new file copyright notice?

lacker replied:
Ah, just copy it exactly from another file in the same directory. So in this case it should be:

/**
 * The examples provided by Facebook are for non-commercial testing and
 * evaluation purposes only.
 *
 * Facebook reserves all rights not expressly granted.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
 * OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NON INFRINGEMENT. IN NO EVENT SHALL
 * FACEBOOK BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN
 * AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

fkgozali commented on Examples/UIExplorer/UIExplorerUnitTests/RCTUnicodeDecodeTests.m:
```diff
@@ -0,0 +1,46 @@
+- (void)testUnicodeDecode
+{
+	NSString* unicodeString = @"שששששששששששששששששששששששששששששששששששששששש";
+	NSData* unicodeBytes = [unicodeString dataUsingEncoding:NSUTF8StringEncoding];
```
nit: could you indent these by 2 spaces throughout this function?

fkgozali commented on Libraries/Network/RCTNetworking.mm:
```diff
@@ -333,22 +333,60 @@ - (RCTURLRequestCancellationBlock)processDataForHTTPQuery:(nullable NSDictionary
 }
 
-+ (NSString *)decodeTextData:(NSData *)data fromResponse:(NSURLResponse *)response
++ (NSString *)decodeTextData:(NSData *)data fromResponse:(NSURLResponse *)response withCarryStorage:(NSMutableData*)carryStorage
```
I think `carryData` is probably suitable variable/method name

fkgozali commented on Libraries/Network/RCTNetworking.mm:
```diff
@@ -333,22 +333,60 @@ - (RCTURLRequestCancellationBlock)processDataForHTTPQuery:(nullable NSDictionary
+			while(carryStorage.length > 0 && newCarryStorage.length < maxCharLength)
+			{
+				[newCarryStorage replaceBytesInRange:NSMakeRange(0, 0) withBytes:[carryStorage subdataWithRange:NSMakeRange(carryStorage.length - 1, 1)].bytes length:1];
+				carryStorage.length -= 1;
```
instead of moving 1 byte for each iteration in the loop, why not use a local variable starting with `carryStorage.length`, then decrement that in the loop, then move the bytes around after the while loop. That's probably more efficient

fkgozali commented on Libraries/Network/RCTNetworking.mm:
```diff
@@ -333,22 +333,60 @@ - (RCTURLRequestCancellationBlock)processDataForHTTPQuery:(nullable NSDictionary
+		}
+	}
+	
+	[carryStorage setData:newCarryStorage];
```
if `carryStorage` was not provided by the caller, this is wasteful.

LeoNatan replied:
@fkgozali This is the original implementation. I opted to use that logic in cases where no carry is provided (such as when non-incremental network request is performed. 

Should be probably `nil` in such cases that the data does not decode correctly with the encoding specified in the header.

fkgozali commented on Libraries/Network/RCTNetworking.mm:
```diff
@@ -333,22 +333,60 @@ - (RCTURLRequestCancellationBlock)processDataForHTTPQuery:(nullable NSDictionary
+	if(carryStorage == nil)
+	{
+		carryStorage = [NSMutableData new];
+	}
```
see my comment below about when `carryStorage` wasn't supplied. It's probably cleaner to make a new working variable and keep the original carryStorage

```
NSMutableData *currentCarry = carryStorage ?: [NSMutableData new];
```

then `if (wasCarryStorageProvided)` can be replaced with `if (carryStorage)` and at the end of the function:

```
if (carryStorage != nil) {
  [carryStorage setData:newCarryStorage];
}
```

This way, it's clearer that you're modifying the supplied data

fkgozali commented on Libraries/Network/RCTNetworking.mm:
```diff
@@ -417,13 +456,23 @@ - (void)sendRequest:(NSURLRequest *)request
+				
+				//Update progress to include the previous carry length and reduce the current carry length.
+				NSArray<id> *responseJSON = @[task.requestID, responseString, @(progress + initialCarryLength - incrementalDataCarryStorage.length), @(total)];
+				
```
a few code styling nits to address:
* comment lines: add space after `//`
* please make sure indentation is 2 spaces consistently
* for long method call, could you split them into multiple lines (e.g. line 474, 467, 368)?
* `if(` --> `if (`, `while(` --> `while (` (space in between)
* curly braces for `if`, `else`, and `while` should start at the same line: `if (...) {`

LeoNatan commented:
@fkgozali @lacker 
Please review revised code with your remarks. Thanks

fkgozali commented on Examples/UIExplorer/UIExplorerUnitTests/RCTUnicodeDecodeTests.m:
```diff
@@ -0,0 +1,59 @@
+
+- (void)testUnicodeDecode
+{
+  NSString* unicodeString = @"😀😀";
```
for safer integration with source control, we usually don't put unicode character directly in the code. Instead we use `\u` representation, so I think this line can be written as:

`@"\uD83D\uDE00\uD83D\uDE00";`

fkgozali commented on Libraries/Network/RCTNetworking.mm:
```diff
@@ -333,22 +333,57 @@ - (RCTURLRequestCancellationBlock)processDataForHTTPQuery:(nullable NSDictionary
+      newCarryData = [currentCarryData subdataWithRange:NSMakeRange(finalLength, currentCarryData.length - finalLength)];
+      currentCarryData.length = finalLength;
+    }
+    else {
```
nit: move this to the same line:

```
} else {
```

fkgozali commented on Examples/UIExplorer/UIExplorerUnitTests/RCTUnicodeDecodeTests.m:
```diff
@@ -0,0 +1,59 @@
+
+@end
+
+@interface RCTUnicodeDecodeTests : XCTestCase @end
```
nit: `@end` should be in a separate line

fkgozali commented on Libraries/Network/RCTNetworking.mm:
```diff
@@ -417,13 +453,28 @@ - (void)sendRequest:(NSURLRequest *)request
+        
+        NSString *responseString = [RCTNetworking decodeTextData:data
+                                                    fromResponse:task.response
+                                                withCarryData:incrementalDataCarry];
```
nit: align the `:` to the previous line

fkgozali commented:
@mkonicek - do you know what went wrong with the travis-ci check?

lacker commented:
<img width="1385" alt="screen shot 2016-10-28 at 9 51 24 am" src="https://cloud.githubusercontent.com/assets/106475/19814688/2ecc20d4-9cf4-11e6-8eef-be6d14b3b64b.png">

Looks like `testUnicodeDecode` failed so it seems like a real failure.

That Travis output is super-confusing - if you just download the raw log, then scroll all the way to the bottom, you get the real error. If you attempt to scroll on the website then you just get 100,000 lines of worthless nonsense ;-)

LeoNatan commented:
@lacker Investigating. Will also incorporate the suggestions. Thanks

LeoNatan commented:
Interesting, the test has exposed an issue I did not think of - Emojis are unicode sequence, so longer than single character. So right now, handling is incorrect if the split comes at a particular location in the string. Investigating a solution.
Unicode parsing is hard!

LeoNatan commented:
@lacker @fkgozali Please see fix and incorporated remarks.

fkgozali commented:
alright, just waiting for the tests to complete

lacker commented:
Now I am extra glad that you added that test! It might be worth testing with a wider array of unusual character strings, as well.

LeoNatan commented:
I think the current implementation covers all unusual sequences. I was not able to find a sequence of three or more characters, and I added a fallback where a longer sequence, if it exists, will be handled correctly.

LeoNatan commented:
Hmm, maybe [Hebrew niqqud](https://en.wikipedia.org/wiki/Niqqud) would fit this "unusual sequence". Will add to test. Also want to test other encodings than UTF8. Maybe other hidden "gems" in there.

LeoNatan commented:
@lacker  @fkgozali 
I am seeing an issue with multibyte encodings, such as UTF16 and UTF32. BOM may or may not be specified, and will only appear at the start of the first data packet. Fixing this issue for multibyte encoding may be more complex than originally thought. 

Right now, native converts the data packet to a string, and then passes it to the JS implementation of XMLHttpRequest for storage (as a string). Any reason why the bytes are not passed directly to JS? Seems to me like this is the most correct solution. Otherwise, a lot of hoops have to be jumped through to handle this correctly.

LeoNatan commented:
I could limit this PR to UTF8, where it works correctly, and is statistically an important fix for most uses cases. However, if the entire unicode spec is to be supported correctly, parsing data at the end, after it has been downloaded sounds the best way.

fkgozali commented:
> Any reason why the bytes are not passed directly to JS

There's a limitation in the current JS <--> native communication mechanism via the bridge where data needs to be serialized (i.e. JSON stringification), which is why things are converted to string.

I'd say to unblock UTF-8 use case, we could limit this PR to UTF-8 (please add comment in the code) then have the UTF16/32 handling in a follow-up

@lacker: thoughts?

LeoNatan commented:
@fkgozali Off topic for this PR, but any reason why bridge uses JSON stringification rather than other, more performant marshaling techniques using JSContext's native bridging?

lacker commented:
I think limiting this fix to UTF-8 sounds like a good short-term workaround. This bug is biting a lot of people and in practice fixing UTF-8 is going to make them happy. Just be clear in the comments that this is what we're doing. And it would be awesome to fix other encodings in a followup PR too <3

LeoNatan commented:
Alright guys, added a limit, added a comment explaining why and added a test with Hebrew niqqud. Used a base64 encoded string to get the Unicode characters in without causing issues with source control systems.

fkgozali commented on Examples/UIExplorer/UIExplorerUnitTests/RCTUnicodeDecodeTests.m:
```diff
@@ -0,0 +1,78 @@
+ */
+
+@import XCTest;
+@import ObjectiveC;
```
it looks like these 2 lines are causing failures internally. could you remove these and replace with

`#import <XCTest/XCTest.h>` ?

LeoNatan commented:
@fkgozali Please try now.

fkgozali commented:
builds seem to be fine internally now, we'll merge this some time monday and let you know if anything else needs attention

fkgozali commented:
OK, this successfully landed internally and github, hence reflected here. We'll be monitoring the internal usage just in case something else breaks due to this change, and if so, will update this thread.

Otherwise, thanks for fixing this issue.

@lacker ^ fyi

LeoNatan commented:
Thank you for the support @fkgozali and @lacker  👍
